Add Uri subclass to avoid dynamic properties

markstory opened the pull request:
Refs https://github.com/cakephp/cakephp/issues/16158

Remove the usage of dynamic properties on Uri by subclassing and making the properties concrete. I'm not sure I like using the builder methods to reconstruct the URI. Another solution I might explore is to create proxy methods, but first I want to profile the runtime impacts of this change.

markstory committed:
Add Uri subclass to avoid dynamic properties

Remove the usage of dynamic properties on Uri by subclassing and making
the properties concrete. I'm not sure I like using the builder methods
to reconstruct the URI. Another solution I might explore is to create
proxy methods, but first I want to profile the runtime impacts of this
change.

othercorey commented on src/Http/Uri.php:
```diff
@@ -0,0 +1,84 @@
+    public function __get($prop): ?string
+    {
+        if ($prop !== 'base' && $prop !== 'webroot') {
+            return null;
```
Is it convention to call trigger_error() here?

ADmad replied:
trigger_error() would allow us to show a better error message.

othercorey replied:
This is used the php docs example:
```php
        trigger_error(
            'Undefined property via __get(): ' . $name .
            ' in ' . $trace[0]['file'] .
            ' on line ' . $trace[0]['line'],
            E_USER_NOTICE);
```

markstory replied:
Yeah I'm not digging how this interface is working. I'm tempted to introduce a container for the Uri & our custom attributes and extract the Uri in the request. My hunch is that very few or no userland code is mucking with the guts of building a request.

othercorey commented:
The memcached error is due to invalid parameter parsing in the extension. It was patched, but they haven't released for 2 years.  

The error went away so I assumed setup-php was building from git, but guess not.

markstory committed:
Use a decorator instead.

I think I like this more than subclassing a library class. The decorator
will also be easier to take out later I think.

markstory commented:
I rejigged this to use a decorator instead of a subclass. I think it will add less overhead than copying all the data into the subclass.

ADmad commented:
LGTM, just the static analyzer errors need to be sorted out.

saeideng commented on src/Http/Uri.php:
```diff
@@ -0,0 +1,219 @@
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.3.0
```
```suggestion
 * @since         4.4.0
```

also in `UriTest.php ` file

markstory committed:
Apply suggestions from code review

Co-authored-by: saeideng <saeideng@users.noreply.github.com>

markstory committed:
Fix linter error.

ADmad commented on src/Http/Uri.php:
```diff
@@ -0,0 +1,219 @@
+    /**
+     * Constructor
+     *
+     * @param \Psr\Http\Message\Uri $uri Uri instance to decorate
```
```suggestion
     * @param \Psr\Http\Message\UriInterface $uri Uri instance to decorate
```

ADmad committed:
Fix type in docblock

markstory committed:
Fix psalm errors

The interface type hints want static returned so we should match that.

markstory committed:
Fix version number.

ADmad commented on src/Http/Uri.php:
```diff
@@ -158,55 +158,55 @@ public function getFragment()
     public function withScheme($scheme)
     {
-        return $this->uri->withScheme($scheme);
+        return new static($this->uri->withScheme($scheme), $this->base, $this->webroot);
```
Swapped one error for another 😆. Static analysers are not easy to please.

markstory committed:
Fix remaining typing issues

new static isn't allowed by phpstan.
